### ISSUE TITLE
Force user to change password on next sign in

### ISSUE THREAD

**Issue by frankieeedeee** [delight-im/PHP-Auth#115]
Hey there,

Just wondering if there's any feature in the library to target a user by their id/email address and force them to change their password on their next sign in? For me, this would allow the users of an application I'm building to create user accounts on behalf of users, with a temporary password, and force the user to change their password on sign in. Should I go about this by keeping track in a separate table referencing their ID's etc?

Love this library so much, thanks for any help!

Cheers

**Comment by ocram** [comment]
Thanks!

This feature is unfortunately not available yet. But we definitely see this as a valid use case. There are other situations where you may also want to force a password reset, e.g. when you detect suspicious activity for a certain user’s account.

So we’d definitely be willing to integrate the feature into this library, but we’re not sure if there’s actually much we can do here.

Forcing the user to reset their password is something that is mainly done in your UI, right? And we can’t control your UI from this library. So all we could do is probably setting a flag in the session variables which you can check constantly so that you know when you need to change your UI to force a password reset. Do you agree?

And if setting such a boolean flag is all we can do, perhaps it’s indeed better to implement this yourself for now, as you suggested. Keeping track of this in a separate table and referencing the users by their IDs should work.

**Comment by frankieeedeee** [comment]
Thanks for the great reply!

Appreciate your help and this library so much, can’t speak highly enough of it

Definitely some other use cases as well, as you mentioned. Potentially even forcing a user to reset their password after x weeks/months in a corporate environment even.

Storing it in the session is one way, however I think it’s a bit beyond the UI alone. Of course, the UI would notify the user that they are required to renew their password, though I see the UI in this context as a representation of state, that state being whether or not the user’s password is current or expired.

Thus, maybe that boolean flag could be an additional column in the users table (“password_expired” perhaps), where 1/true represents that the stored password is no longer current and is due for an update by the user. In the case of this, an attempt on $auth->logIn(...) could throw an exception (ie, “PasswordExpiredException”) giving context of the aforementioned state to the UI, and letting the UI handle from thereonin.

Only a suggestion! But if this column/flag and additional exception existed, it could be framed in a generic way, so that it could fit all the use cases we have both mentioned, and depend on the UI for context as to why the password is expired/no longer current.

Again, only my two cents :) Thanks so much for the great effort!

**Comment by ocram** [comment]
Thanks for your appreciation!

Forcing the user to change their password after a given number of weeks or months in a corporate environment would certainly be another use case, to the extend that you deem that process a reasonable practice.

Storing the state of the password expiry in the session and exposing access to that information via an accessor method such as `Auth#hasPasswordExpired` would allow you to both change your UI accordingly and guard calls to any other methods from this library. That would definitely allow for a secure implementation.

Throwing exceptions from every single method when the password has expired would indeed be possible as well, though I’m not quite sure yet that this is something we’d want to do. Again, it’s possible, but keeping this library easy and comfortable to use for developers, we have to think about this more carefully.

Storing the information in the `users` table in a column named `password_expired`, for example, is definitely the way we’d do this.

All this being said, we can’t promise that this feature will be implemented soon. We’re not sure about the exceptions, and everything else can easily be implemented on top of this library in your own code, for now.

**Comment by frankieeedeee** [comment]
Absolutely! Totally understand - We’ll keep it to an individual inplementation for now, and I’ll keep my ear to the ground for any updates in the future with the above features

Cheers 👍

**Comment by kennyrkun** [comment]
Would a `password_expire_date` column where the value is either null for no expiration, a date in the past for an expired password, or a date in the future for a valid password be an optimal solution? Null being default would allow for use like normal. When a webmaster decides, for whatever reason, that a password should be reset, they can simply set the expiration date to the current date or any date in the past, and it will be treated as an expired password. A date in the future would allow for easy automatic password expiration.

The library need simply to throw a `PasswordExpired` during initial `$auth->login()` and then in every call to `$auth->check()` or `$auth->isLoggedIn()`. Not immediately logging the user out for an expired password would not be dissimilar from the delay experienced in calls like `$auth->logOutEverywhere()` or `$auth->logOutEverywhereElse()`.

**Comment by kennyrkun** [comment]
Another possible option is a column containing the last date the password was updated. Applications could then apply their own logic based on that date. That requires minimal work and imposes minimal restrictions from the part of the library.

**Comment by ocram** [comment]
The new major version is coming with an audit log that tracks actions relevant for the security of users’ accounts, including password changes and password resets. All data is stored in a new table called `users_audit_log`.

So if you only want to find out when the user last changed their password, to decide yourself whether that is fine or not, and whether the password should be considered as expired or not, then you will probably be able to just look it up using the table from the audit log:

```sql
SELECT MAX(event_at) AS password_last_changed_timestamp FROM users_audit_log WHERE user_id = ? AND (event_type = 'password.change' OR event_type = 'password.reset.finish');
```

If that’s not what you want, then the recommendation is still to just create your own simple table in the database that tracks whether users need to change their passwords or not, e.g. tracking `(user_id, password_expired_bool)` or `(user_id, password_expired_from_timestamp)`.

In any case, you can decide for yourself what restrictions to put into the UI (and perhaps backend logic) for users.

We’ll keep this issue here open until we find out whether there is a better way to force users to change their password right from the library and perhaps without any custom code.
